File: app/src/main/java/com/shrikantbadwaik/tmdb/view/base/BaseViewModel.java
package com.shrikantbadwaik.tmdb.view.base;

import android.arch.lifecycle.ViewModel;
import android.content.Intent;

import com.shrikantbadwaik.tmdb.data.repository.Repository;

import java.lang.ref.WeakReference;

public abstract class BaseViewModel<V extends BaseView> extends ViewModel {
    private final Repository repository;
    private WeakReference<V> view;

    public BaseViewModel(Repository repository) {
        this.repository = repository;
    }

    @Override
    protected void onCleared() {
        view.clear();
    }

    public void attachView(V view) {
        this.view = new WeakReference<>(view);
    }

    public boolean isViewAttached() {
        return view != null;
    }

    public V getView() {
        return view.get();
    }

    public Repository repository() {
        return repository;
    }

    public void onActivityStarted() {
    }

    public void onActivityResumed() {
    }

    public void onActivityPaused() {
    }

    public void onActivityStopped() {
    }

    public void onActivityDestroyed() {
    }

    public void onActivityResult(int requestCode, int resultCode, Intent data) {
    }

    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
    }
}

File: app/src/main/java/com/shrikantbadwaik/tmdb/view/base/BaseViewModel.kt
package com.shrikantbadwaik.tmdb.view.base

import android.arch.lifecycle.ViewModel
import android.databinding.ObservableBoolean
import com.shrikantbadwaik.tmdb.data.repository.Repository
import java.lang.ref.WeakReference


abstract class BaseViewModel<View : BaseView>(private val repository: Repository) : ViewModel() {
    private lateinit var view: WeakReference<View>
    private val loading: ObservableBoolean = ObservableBoolean()

    fun attachView(view: View) {
        this.view = WeakReference(view)
    }

    fun getView(): View? {
        return view.get()
    }

    fun repository(): Repository {
        return repository
    }

    fun setLoading(loading: Boolean) {
        this.loading.set(loading)
    }

    fun isLoading() = loading

    open fun onActivityStarted() {}

    open fun onActivityResumed() {}

    open fun onActivityPaused() {}

    open fun onActivityStopped() {}

    open fun onActivityDestroyed() {}

    override fun onCleared() {
        super.onCleared()
        view.clear()
    }
}